interceptor/tracing_interceptor.go

```diff
@@ -0,0 +1,476 @@
+// The MIT License
+//
+// Copyright (c) 2021 Temporal Technologies Inc.  All rights reserved.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package interceptor
+
+import (
+	"context"
+	"fmt"
+
+	commonpb "go.temporal.io/api/common/v1"
+	"go.temporal.io/sdk/activity"
+	"go.temporal.io/sdk/client"
+	"go.temporal.io/sdk/converter"
+	"go.temporal.io/sdk/workflow"
+)
+
+const (
+	workflowIDTagKey = "temporalWorkflowID"
+	runIDTagKey      = "temporalRunID"
+)
+
+// Tracer is an interface for tracing implementations as used by
+// NewTracingInterceptor. Most callers do not use this directly, but rather use
+// the opentracing or opentelemetry packages.
+type Tracer interface {
+	// SpanContextKey provides a key to put a span on a context unrelated to how a
+	// span might otherwise be put on a context by ContextWithSpan. This should
+	// never be nil.
+	SpanContextKey() interface{}
```

Comment:
After use of this in another PR, it has become this should be `Options() TracerOptions` that is invoked one time that has options instead of `SpanContextKey` and `HeaderKey` separate.

Reply:
In typescript, SpanContext key is tied to the baggage and tracing propagators.

Reply:
Span context key relates to go contexts, and is tied to the opentracing and opentelemetry tracers.